[v2_lsqmr] Implement LSQR and LSMR solvers and their customized Stopping Criterion.

okumuskaan opened the pull request:
* Theory behind LSQR and LSMR and their implementation is studied by Stanford articles: [LSQR](https://web.stanford.edu/group/SOL/software/lsqr/lsqr-toms82a.pdf) and [LSMR](https://web.stanford.edu/group/SOL/software/lsmr/LSMR-SISC-2011.pdf). All estimation methods, stopping criterions are implemented from these techniques. 
* `LSQR` and `LSMR` classes are implemented in `lsqr.py` and `lsmr.py` with the inspiration from [scipy.sparse.linalg.lsqr](https://github.com/scipy/scipy/blob/v1.8.1/scipy/sparse/linalg/_isolve/lsqr.py#L96-L586) and [scipy.sparse.linalg.lsmr](https://github.com/scipy/scipy/blob/v1.8.1/scipy/sparse/linalg/_isolve/lsmr.py#L29-L485).
* `StopCriterion_LSQMR` class is implemented in `stop.py`. This is the customized stopping criterion to work with LSQR and LSMR.
* `LSQR` and `LSMR` work with multiple inputs and the solution is obtained in parallel.
* Documentation for each solvers and stopping criterion are done.

okumuskaan committed:
[opt/stop] Implement stopping criterions for lsqr/lsmr.

okumuskaan committed:
[opt/solver/lsqr] Create and implement lsqr.

okumuskaan committed:
[opt/solver/lsmr] Create and implement lsmr.

okumuskaan committed:
[opt/solver/lsqr] Document and update lsqr.py

- Final documentation is done.
- The function is made to work for multiple inputs by parallel processing.
- Stopping criterion is updated.
- The function is made to respond to trivial solutions.

okumuskaan committed:
[opt/solver/lsmr] Document and update lsmr.py

- Final documentation is done.
- The function is made to work for multiple inputs by parallel processing.
- Stopping criterion is updated.
- The function is made to respond to trivial solutions.

okumuskaan committed:
[opt] Document and implement StopCriterion_LSQMR in stop.py

- Customized stopping criterion is implemented to be used for lsqr.py and lsmr.py.
- Instead of creating two different classes, one generic class is implemented.
- Documentation is finalized.

okumuskaan committed:
[doc] Documentation is updated in pycsou.opt.solver.rst.

okumuskaan committed:
[opt/solver/lsqr] Reference is updated.

joanrue commented on src/pycsou/opt/solver/lsmr.py:
```diff
@@ -0,0 +1,332 @@
+        verbosity: int = 1,
+        show_progress: bool = True,
+        log_var: pyct.VarName = ("x",),
+    ):
```
Use kwargs to avoid cascading Solver.init() changes. 
```
    def __init__(
        self,
        A: pyco.PosDefOp,
        *,
        damp: float = 0.0,
        atol: float = 1e-06,
        btol: float = 1e-06,
        conlim: float = 1e08,
        iter_lim: typ.Optional[int] = None,
        **kwargs
    ):
        super().__init__(**kwargs)
```

okumuskaan replied:
I fixed that !

joanrue commented on src/pycsou/opt/solver/lsmr.py:
```diff
@@ -0,0 +1,332 @@
+         (..., N) 'b' terms in the LSMR cost function. All problems are solved in parallel.
+    x0: NDArray
+        (..., N) initial point(s). Defaults to 0 if unspecified.
+    restart_rate: int
```
I think the `restart_rate` parameter is described in the documentation but is not present in the implementation.

okumuskaan replied:
Fixed that.

joanrue commented on src/pycsou/opt/solver/lsmr.py:
```diff
@@ -0,0 +1,332 @@
+        m, n = self._A.shape
+
+        u = b
+        normb = xp.linalg.norm(b)
```
Seems that beta is the norm in the dimension of interest (not the stacking dimensions). If this indeed the case, then  `normb = xp.linalg.norm(b, axis=-1, keepdims=True)` should do it.

okumuskaan replied:
Fixed that.

joanrue commented on src/pycsou/opt/solver/lsmr.py:
```diff
@@ -0,0 +1,332 @@
+        else:
+            x = x0.copy()
+            u -= self._A.apply(x)
+            beta = xp.linalg.norm(u)
```
same as above

okumuskaan replied:
Fixed that.

joanrue commented on src/pycsou/opt/solver/lsmr.py:
```diff
@@ -0,0 +1,332 @@
+            u -= self._A.apply(x)
+            beta = xp.linalg.norm(u)
+
+        if beta > 0:
```
given beta can be NDArray this needs to be handled differently. Maybe something like
`alpha = (beta > 0) * xp.linalg.norm(v, axis=-1, keepdims=True)`

okumuskaan replied:
Fixed that.

joanrue commented on src/pycsou/opt/solver/lsmr.py:
```diff
@@ -0,0 +1,332 @@
+        if beta > 0:
+            u *= 1 / beta
+            v = self._A.T.apply(u)
+            alpha = xp.linalg.norm(v)
```
same as with beta?

okumuskaan replied:
Fixed that.

joanrue commented on src/pycsou/opt/solver/lsmr.py:
```diff
@@ -0,0 +1,332 @@
+                mst["x"] = xp.zeros(n)
+            elif b.ndim == 2:
+                mst["x"] = xp.zeros((b.shape[0], n))
+            return
```
need to handle case of b.ndim>2

okumuskaan replied:
Fixed that by using xp.zeros_like.

joanrue commented on src/pycsou/opt/solver/lsmr.py:
```diff
@@ -0,0 +1,332 @@
+
+        # Bidiagonalizaion to obtain next beta, u, alpha, v:
+        u = self._A.apply(v) - alpha * u
+        beta = xp.linalg.norm(u)
```
same as above

okumuskaan replied:
Fixed that.

joanrue commented on src/pycsou/opt/solver/lsmr.py:
```diff
@@ -0,0 +1,332 @@
+        mst["test1"], mst["test2"], mst["test3"] = 1, alpha / beta, None
+
+    @staticmethod
+    def _sym_ortho(a, b, xp):
```
`a` and `b` can be NDArrays, would it make sense to rotate each batching element?

okumuskaan replied:
Fixed that by using xp.vectorize function. This is the only function that exists both in numpy, dask.array and cupy. I am not sure if it's the fastest solution to make this function to be vectorized, but it works with that. There is also frompyfunc function, but it doesn't exist in cupy, that's why I used vectorize

joanrue commented on src/pycsou/opt/solver/lsmr.py:
```diff
@@ -0,0 +1,332 @@
+        rhodold, zeta, d = mst["rhodold"], mst["zeta"], mst["d"]
+        normA2, minrbar, maxrbar = mst["normA2"], mst["minrbar"], mst["maxrbar"]
+
+        self._itn += 1
```
Would it be useful to use the solver iteration counter `self._astate["idx"]`?

okumuskaan replied:
Fixed that.

joanrue commented on src/pycsou/opt/solver/lsqr.py:
```diff
@@ -0,0 +1,319 @@
+         (..., N) 'b' terms in the LSQR cost function. All problems are solved in parallel.
+    x0: NDArray
+        (..., N) initial point(s). Defaults to 0 if unspecified.
+    restart_rate: int
```
same comments from LSMR apply for LSQR

okumuskaan replied:
Done.

joanrue commented on src/pycsou/opt/solver/lsqr.py:
```diff
@@ -0,0 +1,319 @@
+        mst["test3"] = mst["t1"] = mst["rtol"] = None
+
+    @staticmethod
+    def _sym_ortho(a, b, xp):
```
would it be interesting to use a base class for LSMR and LSQR which defines the common methods, _sym_ortho, solution, input handling, etc..?

matthieumeo replied:
Haven't had a look at the code, but it could be a good idea indeed: the two algorithms are very similar and if there is redundancy, let's centralise it into a private class (maybe even an abstract one as _PDS).

joanrue reviewed (commented):
Amazing job! the main comments are related to allowing `b` to have more than 2 dims.

joanrue reviewed (commented):
Amazing job! The main comments are related to allowing `b` to have more than 2 dims.

okumuskaan committed:
[opt/solver/lsqr] Update lsqr.py according to Joan's review

- The code is updated to work for multiple i/o system in parallel. In the old work, it's detected that the code doesn't work for multiple inputs. In the updated version, the solver continues the iteration until all tests are passed. Tests for each multiple is done in parallel.
- The solver is updated such that it can deal with trivial solution and doesn't do unnecessary computation for the inputs that have trivial solution.

okumuskaan committed:
[opt/solver/lsmr] Update lsmr.py according to Joan's review

- The code is updated to work for multiple i/o system in parallel. In the old work, it's detected that the code doesn't work for multiple inputs. In the updated version, the solver continues the iteration until all tests are passed. Tests for each multiple is done in parallel.
- The solver is updated such that it can deal with trivial solution and doesn't do unnecessary computation for the inputs that have trivial solution.

okumuskaan committed:
[opt/stop] Update StopCriterion_LSQMR according to latest changes

- It is updated to deal with multiple i/o systems working in parallel.
- Tests for each inputs/outputs are done accordingly.

okumuskaan commented:
There are some updates that I did according to Joan's review.
* kwargs is used to avoid cascading Solver.init() changes for the initialization of `LSQR` and `LSMR`.
* Mistakenly written restart_rate parameter in the documentation of `LSQR` and `LSMR`  is removed.
* `LSQR`, `LSMR` and `StopCriterion_LSQMR` are updated to work for multiple inputs/outputs. There were major errors in the code. One of them is that testing is not done in parallel for each i/o. This is fixed by changing the most of the code. Now, the testings for each i/o is done and the solver stops whenever the testings for every i/o is passed. In order not to display the whole data, I decided to display the info only for one i/o system. 
* There might be trivial solutions, where output is full of zeros, so the input should be full of zeros too without solving anything. After the recent updates about multiple i/o implementation, trivial solution should be taken into consideration. If all of the i/o has trivial solution, it will be easy to say that all the inputs are full of zero. But in the cases where some i/o parts have trivial solution while other parts do not, we need to solve only nontrivial i/o's but have the correct solution for whole system. This is done by detecting which i/o's have trivial solution and remove them from the i/o's for the solving process such that only nontrivial ones will be solved. At the end of the solving process, trivial solutions are added to nontrivial solution accordingly. This is the best solution that came to my mind, if you have any other suggestions, please let me know.

joanrue commented on src/pycsou/opt/solver/lsqr.py:
```diff
@@ -0,0 +1,311 @@
+        mst["test3"] = mst["t1"] = mst["rtol"] = None
+
+        # Vectorize the function of Givens rotation
+        self._sym_ortho_func = xp.vectorize(self._sym_ortho)
```
Dask and Cupy aren't currently supporting np.vectorize

joanrue commented on src/pycsou/opt/solver/lsqr.py:
```diff
@@ -0,0 +1,311 @@
+    >>> lsqr = LSQR(A, show_progress=False)
+    >>> lsqr.fit(b=b)
+    >>> x_solution = lsqr.solution()
+    >>> assert np.allclose(x_star, x_solution)
```
Maybe add  some tolerance to not get an error (e.g., `assert np.allclose(x_star, x_solution, rtol=1e-2)`)

joanrue commented on src/pycsou/opt/solver/lsqr.py:
```diff
@@ -0,0 +1,311 @@
+        mst["alpha"], mst["rhobar"], mst["phibar"] = alpha, rhobar, phibar
+
+        # Estimation of norms:
+        ddnorm = ddnorm + xp.linalg.norm(dk) ** 2
```
which axis should the norm be applied? Dask will only work for 1D or 2D inputs when axis is not specified.
